addons/web/static/src/core/virtual_hook.js

```diff
@@ -0,0 +1,119 @@
+/** @odoo-module **/
+
+import { onWillRender, onWillStart, toRaw, useEffect, useState } from "@odoo/owl";
+import { shallowEqual } from "@web/core/utils/arrays";
+import { throttleForAnimation } from "@web/core/utils/timing";
+
+/**
+ * @template T
+ * @typedef VirtualHookParams
+ * @property {T[] | () => T[]} items
+ * @property {typeof useRef} scrollableRef
+ * @property {ScrollPosition} [initialScroll={ top: 0 }]
+ * @property {PixelValue | (item: T) => PixelValue} [itemHeight=0]
+ * @property {PixelValue} [margin="100%"]
+ */
+
+/**
+ * @typedef ScrollPosition
+ * @property {number} top
+ */
+
+/** @typedef {number | `${string}px` | `${string}%`} PixelValue */
+
+/**
+ * Converts a number,
+ *
+ * @param {PixelValue} value
+ * @returns {number}
+ */
+const toPixels = (value) => {
+    if (typeof value === "number") {
+        return value;
+    }
+    if (value.endsWith("%")) {
+        return window.innerHeight * (Number(value.slice(0, -1)) / 100);
+    }
+    if (value.endsWith("px")) {
+        return Number(value.slice(0, -2));
+    }
+};
+
+/**
+ * Calculates the displayed items in a virtual list.
+ *
+ * @template T
+ * @param {VirtualHookParams<T>} params
+ * @returns {ReturnType<useState<T>>}
+ */
+export function useVirtual({ items, scrollableRef, initialScroll, itemHeight, margin }) {
```

Comment:
looks like everything is ready to enable the x axis virtualization as well, maybe except for the margin param. Maybe consider naming it yMargin

Reply:
Ok I just understood what margin stands for. My previous comment can probably be ignored then.
Not sure *margin* is the best name though.

Reply:
Cool hook. Just wondering whether it should also deal with the blank areas itself or not. As far as i understand, it's the caller of the hook that has to set the correct height on the virtualized container and (I guess) to add blank elements before and after the visible items.

Reply:
There are no "blank spacers" required.
The caller must set specific styling rules in order to make this work.
As it is stated in the commit message, a styling strategy is to use display-grid.

We should update the commit message as it is not the only styling strategy possible.
The test suite make use of computed relative positioning (top: Xpx)

Reply:
The commit message has been updated to state the hook's requirements to put in place.

Also, some simplifications:
- the `margin` option is removed, allowing the deletion of the local utility function `toPixels(value: PixelValue)` (the type `PixelValue` is removed too)
- `items` and `itemHeight` are renamed `getItems` and `getItemHeight` (they now are always functions)